File: dolfin/la/KrylovSolver.cpp
// Copyright (C) 2007-2010 Garth N. Wells
//
// This file is part of DOLFIN.
//
// DOLFIN is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// DOLFIN is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with DOLFIN. If not, see <http://www.gnu.org/licenses/>.
//
// Modified by Ola Skavhaug 2008
// Modified by Anders Logg 2008-2012
//
// First added:  2007-07-03
// Last changed: 2013-11-25

#include <dolfin/common/Timer.h>
#include <dolfin/parameter/GlobalParameters.h>
#include <dolfin/parameter/Parameters.h>
#include <dolfin/log/LogStream.h>
#include "GenericMatrix.h"
#include "GenericVector.h"
#include "DefaultFactory.h"
#include "LinearSolver.h"
#include "VectorSpaceBasis.h"
#include "KrylovSolver.h"

using namespace dolfin;

//-----------------------------------------------------------------------------
Parameters KrylovSolver::default_parameters()
{
  Parameters p("krylov_solver");

  p.add<double>("relative_tolerance");
  p.add<double>("absolute_tolerance");
  p.add<double>("divergence_limit");
  p.add<int>("maximum_iterations");
  p.add<bool>("report");
  p.add<bool>("monitor_convergence");
  p.add<bool>("error_on_nonconvergence");
  p.add<bool>("nonzero_initial_guess");

  return p;
}
//-----------------------------------------------------------------------------
KrylovSolver::KrylovSolver(MPI_Comm comm, std::string method, std::string preconditioner)
{
  // Initialize solver
  init(method, preconditioner, comm);
}
//-----------------------------------------------------------------------------
KrylovSolver::KrylovSolver(std::string method, std::string preconditioner)
{
  // Initialize solver
  init(method, preconditioner, MPI_COMM_WORLD);
}
//-----------------------------------------------------------------------------
KrylovSolver::KrylovSolver(MPI_Comm comm,
                           std::shared_ptr<const GenericLinearOperator> A,
                           std::string method, std::string preconditioner)
{
  // Initialize solver
  init(method, preconditioner, comm);

  // Set operator
  set_operator(A);
}
//-----------------------------------------------------------------------------
KrylovSolver::~KrylovSolver()
{
  // Do nothing
}
//-----------------------------------------------------------------------------
void
KrylovSolver::set_operator(std::shared_ptr<const GenericLinearOperator> A)
{
  dolfin_assert(solver);
  solver->parameters.update(parameters);
  solver->set_operator(A);
}
//-----------------------------------------------------------------------------
void
KrylovSolver::set_operators(std::shared_ptr<const GenericLinearOperator> A,
                            std::shared_ptr<const GenericLinearOperator> P)
{
  dolfin_assert(solver);
  solver->parameters.update(parameters);
  solver->set_operators(A, P);
}
//-----------------------------------------------------------------------------
std::size_t KrylovSolver::solve(GenericVector& x, const GenericVector& b)
{
  dolfin_assert(solver);
  //check_dimensions(solver->get_operator(), x, b);

  Timer timer("Krylov solver");
  solver->parameters.update(parameters);
  return solver->solve(x, b);
}
//-----------------------------------------------------------------------------
std::size_t KrylovSolver::solve(const GenericLinearOperator& A,
                                GenericVector& x,
                                const GenericVector& b)
{
  dolfin_assert(solver);
  //check_dimensions(A, x, b);

  Timer timer("Krylov solver");
  solver->parameters.update(parameters);
  return solver->solve(A, x, b);
}
//-----------------------------------------------------------------------------
void KrylovSolver::init(std::string method, std::string preconditioner, MPI_Comm comm)
{
  // Get default linear algebra factory
  DefaultFactory factory;

  // Get list of available methods and preconditioners
  const std::map<std::string, std::string>
    methods = factory.krylov_solver_methods();
  const std::map<std::string, std::string>
    preconditioners = factory.krylov_solver_preconditioners();

  // Check that method is available
  if (!LinearSolver::in_list(method, methods))
  {
    dolfin_error("KrylovSolver.cpp",
                 "solve linear system using Krylov iteration",
                 "Unknown Krylov method \"%s\". "
                 "Use list_krylov_solver_methods() to list available Krylov methods",
                 method.c_str());
  }

  // Check that preconditioner is available
  if (!LinearSolver::in_list(preconditioner, preconditioners))
  {
    dolfin_error("KrylovSolver.cpp",
                 "solve linear system using Krylov iteration",
                 "Unknown preconditioner \"%s\". "
                 "Use list_krylov_solver_preconditioners() to list available preconditioners()",
                 preconditioner.c_str());
  }

  // Set default parameters
  parameters = dolfin::parameters("krylov_solver");

  // Initialize solver
  solver = factory.create_krylov_solver(comm, method, preconditioner);
  solver->parameters.update(parameters);
}
//-----------------------------------------------------------------------------

File: test/unit/la/python/KrylovSolver.py
"""Unit tests for the KrylovSolver interface"""

# Copyright (C) 2012 Johan Hake
#
# This file is part of DOLFIN.
#
# DOLFIN is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# DOLFIN is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with DOLFIN. If not, see <http://www.gnu.org/licenses/>.
#
# Modified by Anders Logg 2012
#
# First added:  2012-02-21
# Last changed: 2012-08-27

import unittest
from dolfin import *

# Assemble system
mesh = UnitSquareMesh(32, 32)
V = FunctionSpace(mesh, 'CG', 1)
bc = DirichletBC(V, Constant(0.0), lambda x, on_boundary: on_boundary)
u = TrialFunction(V); v = TestFunction(V);
u1 = Function(V)

# Forms
a, L = inner(grad(u), grad(v))*dx, Constant(1.0)*v*dx
a_L = inner(grad(u1), grad(v))*dx

# Assemble linear algebra objects
A = assemble(a)
b = assemble(L)
bc.apply(A, b)

if has_linear_algebra_backend("PETSc"):
    class PETScKrylovSolverTester(unittest.TestCase):

        def test_krylov_solver(self):
            "Test PETScKrylovSolver"
            # Get solution vector
            tmp = Function(V)
            x = tmp.vector()

            # Solve first using direct solver
            solve(A, x, b, "lu")

            direct_norm = x.norm("l2")

            # Get solution vector
            x_petsc = as_backend_type(x)

            for prec, descr in krylov_solver_preconditioners():
                if MPI.num_processes() > 1 and prec in ["ilu", "icc", "jacobi", "hypre_amg"]:
                    print "FIXME: Preconditioner '%s' does not work in parallel,"\
                          " skipping" % prec
                    continue

                # With simple interface
                solver = PETScKrylovSolver("gmres", prec)
                solver.solve(A, x_petsc, as_backend_type(b))
                self.assertAlmostEqual(x_petsc.norm("l2"), direct_norm, 5)


                # With PETScPreconditioner interface
                solver = PETScKrylovSolver("gmres", PETScPreconditioner(prec))
                solver.solve(A, x_petsc, as_backend_type(b))
                self.assertAlmostEqual(x_petsc.norm("l2"), direct_norm, 5)

if __name__ == "__main__":

    # Turn off DOLFIN output
    set_log_active(True)

    print ""
    print "Testing DOLFIN la/KrylovSolver interface"
    print "----------------------------------------"
    unittest.main()
